Add mmt4d ukernel for riscv64's zvfhmin and zvfh feature, for types f16xf16->f16/f32

adeel10x opened the pull request:
This PR adds mmt4d ukernel for riscv64's **zvfhmin and zvfh** feature, for types **f16xf16->f16 and f16xf16->f32**. Tests and benchmarks are also enabled for the added ukernels.

`getauxval` call is replaced by `__riscv_hwprobe` call to determine supported riscv hardware extensions at runtime. See: https://github.com/torvalds/linux/blob/master/Documentation/arch/riscv/hwprobe.rst

adeel10x committed:
Add mmt4d ukernel implementation for riscv64 target(zvfhmin feature), for types f16xf16->fXX, using tile Size {M0,N0,K0} = {1,XX,1} to {6,XX,1}

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

adeel10x committed:
Apply pre-commit hooks

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

benvanik commented on runtime/src/iree/base/internal/cpu.c:
```diff
@@ -375,15 +375,35 @@ static void iree_cpu_initialize_from_platform_x86_64(uint64_t* out_fields) {
-#define IREE_HWCAP_ISA_V (1 << ('V' - 'A'))
+// NOTE: not all kernel versions have all of the cap bits we need defined so as
+// a practice we always define the feature bits we need locally.
+#define RISCV_HWPROBE_IMA_V (1 << 2)
```
Prefix with IREE_ or #if !defined() check so we don't collide with kernels who do.

benvanik reviewed (dismissed):
Minor comment but otherwise looks good for the non-ukernel itself (I'll let @bjacob look at that :)

bjacob reviewed (approved):
All looks good, other than the hwproble compilation issue that is caught by CI anyway.

adeel10x committed:
Use riscv_hwprobe syscall directly to avoid dependency on GLIBCv2.40, and add prefix 'IREE' to riscv feature bit names

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

adeel10x committed:
Define required structs locally if <asm/hwprobe.h> header is not available. Define all required constants locally as well since not all kernel versions have all of the constants needed.

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

adeel10x committed:
Define __NR_riscv_hwprobe locally to avoid build failure if it's not defined in <unistd.h> header

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

adeel10x committed:
Add cmake check to verify that the RISC-V toolchain can compile code containing __extendhfsf2 builtin

Add CMake check to verify that the system RISC-V toolchain can successfully link code that has __extendhfsf2 builtin for fp16 to fp32 conversions. This detects cases where the system toolchain emits a call to the `__extendhfsf2` builtin but fails to link it correctly. In such cases corresponding ukernel code compilation is skipped to avoid build failures.

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

adeel10x committed:
Prevent compiler optimization in CMake test for zvfh/zvfhmin feature by using volatile keyword

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>

bjacob commented on runtime/src/iree/builtins/ukernel/arch/riscv_64/CMakeLists.txt:
```diff
@@ -118,8 +118,8 @@ if(IREE_UK_TRY_RISCV_64_ZVFHMIN_ZVFH)
-"  _Float16 val_f16 = 100.0;"
-"  float val_f32 = (float)val_f16;"
+"  volatile _Float16 val_f16 = 100.0;"
+"  volatile float val_f32 = (float)val_f16;"
```
Use of `volatile` to prevent compiler optimation should technically be fine, but I always wonder about local volatile variables if the compiler could be tempted to prove by escape-analysis that the accesses are still fine to optimize away.  Anyway, I wonder if we could do something even simpler. How about performing the f16->f32 on a variable that isn't a constant in the first place ? E.g. could this testcase be

```c
float foo(_Float16 x) { return x; }
```

If it complains about the lack of `main` function, maybe add `-c` ? Or perform the cast on `(_Float16)argc`.

adeel10x replied:
Casting argc to _Float16 (or float afterwards) doesn't result in a call to any runtime builtin.
 For `float foo(_Float16 x) { return x; }`, the compiler complains about the missing main, but I can't use -c to compile this cmake test because it would prevent linking (`undefined reference to __extendhfsf2` is a linking error).

bjacob replied:
I see, thanks for trying anyway! OK to keep the current test as-is.

I see that the tests have passed. Do you want me to merge this for you?

Also consider getting write access!

adeel10x replied:
I'll add zvfh ukernels very soon, and then you can merge this. Yes, write access would be great :).

bjacob replied:
@ScottTodd , how do we give @adeel10x write access?

ScottTodd replied:
https://iree.dev/developers/general/contributing/#obtaining-commit-access

adeel10x committed:
Add mmt4d ukernel implementation for riscv64 target(zvfh feature), for types f16xf16->f16/f32

Signed-off-by: adeel10x <adeel.ahmad@10xengineers.ai>
